### PR TITLE
Add  a new Condition for BEB Webhook callback status

### PR DESCRIPTION
**Pull request by radufa**
**Description**

Changes proposed in this pull request:

- Add  a new condition for BEB Webhook callback status
- The status "ready" of a BEB subscription checks now the status of the new condition too
- Avoid during the reconciler call to work with an outdated subscription object

**Related issue(s)**

### CODE REVIEW

File: components/eventing-controller/controllers/subscription/beb/reconciler.go [hunk 1]
```diff
@@ -643,18 +657,25 @@ func (r *Reconciler) syncInitialStatus(ctx context.Context, subscription *eventi
 	}
 	// case: conditions are already initialized
 	if len(currentStatus.Conditions) >= len(expectedStatus.Conditions) && !updateReadyStatus {
```
Comment (k15r) [inline components/eventing-controller/controllers/subscription/beb/reconciler.go:659]: is that even a valid check? if `currentStatus.Conditions  == [a,b,c]` and `expectedStatus.Conditions == [d,e]` that would pass, but clearly expected conditions would not be in currentStatus

how about always processing the conditions and not return early here?

File: components/eventing-controller/controllers/subscription/beb/reconciler.go [hunk 2]
```diff
@@ -852,6 +873,28 @@ func (r *Reconciler) checkStatusActive(subscription *eventingv1alpha1.Subscripti
 	return false, retry, err
 }
 
+// checkLastFailedDelivery checks if LastFailedDelivery exists and if happened after LastSuccessfulDelivery
+func (r *Reconciler) checkLastFailedDelivery(subscription *eventingv1alpha1.Subscription) bool {
+	if len(subscription.Status.EmsSubscriptionStatus.LastFailedDelivery) > 0 {
+		var lastFailedDeliveryTime, LastSuccessfulDeliveryTime time.Time
+		var err error
+		if lastFailedDeliveryTime, err = time.Parse(time.RFC3339, subscription.Status.EmsSubscriptionStatus.LastFailedDelivery); err != nil {
+			r.namedLogger().Errorw("parse LastFailedDelivery failed", "error", err)
```
Comment (k15r) [inline components/eventing-controller/controllers/subscription/beb/reconciler.go:882]: should we really hide this error here and assume everything is alright, by returning false?

File: components/eventing-controller/api/v1alpha1/subscription_lifecycle.go [hunk 3]
```diff
@@ -36,6 +37,7 @@ const (
 	ConditionReasonAPIRuleStatusReady         ConditionReason = "APIRule status ready"
 	ConditionReasonAPIRuleStatusNotReady      ConditionReason = "APIRule status not ready"
 	ConditionReasonNATSSubscriptionActive     ConditionReason = "NATS Subscription active"
+	ConditionReasonSubscriptionWebhookCall    ConditionReason = "BEB Subscription webhook call"
```
[the remark this answers is not in the document]
Comment (radufa) [reply components/eventing-controller/api/v1alpha1/subscription_lifecycle.go]: changed

File: components/eventing-controller/controllers/subscription/beb/reconciler.go [hunk 4]
```diff
@@ -643,18 +657,25 @@ func (r *Reconciler) syncInitialStatus(ctx context.Context, subscription *eventi
 	}
 	// case: conditions are already initialized
 	if len(currentStatus.Conditions) >= len(expectedStatus.Conditions) && !updateReadyStatus {
-		return nil
+		return false
+	} else if len(currentStatus.Conditions) > 0 {
```
[the remark this answers is not in the document]
Comment (radufa) [reply components/eventing-controller/controllers/subscription/beb/reconciler.go]: true

File: components/eventing-controller/controllers/subscription/beb/reconciler.go (hunk 1 above)
Comment (radufa) [reply components/eventing-controller/controllers/subscription/beb/reconciler.go]: This was the old behavior, I didn't want to change it.
Changed now.

File: components/eventing-controller/controllers/subscription/beb/reconciler.go (hunk 2 above)
Comment (radufa) [reply components/eventing-controller/controllers/subscription/beb/reconciler.go:901]: This error will be reflected in the subscription status too. It's an indicator that we have to update the "Webhook call status" condition to reflect the changes in BEB

**Review by k15r — APPROVED** [review]

**Pull request merged**